File: ray.py
from point import Point
from vector import Vector


class Ray:
    def __init__(self, origin, direction):
        self.origin = origin
        self.direction = direction.normalize()

    def at(self, t):
        return self.origin + t * self.direction

    def __call__(self, t):
        return self.at(t)

File: ray.cpp
#include "ray.h"

Ray::Ray(const Vector &origin, const Vector &direction) : origin(origin), direction(direction) {
    direction.normalize();
}

Vector Ray::at(double t) const {
    return origin + direction * t;
}

Vector Ray::operator()(double t) const {
    return at(t);
}